Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,719 @@
+        num_indices: int
+            The sampler will select from a range of indices 0 to num_indices
+
+        prob: list of floats, optional 
+            The probability for each index to be selected by the 'next' operation. If not provided, the indices will be sampled uniformly. The list should have a length equal to num_indices, and the values should sum to 1
+
+        seed:int, optional
+            Used to initialise the random number generator where repeatability is required.
+
+        Returns  
+        -------  
+        `RandomSampler`  
+            An instance of the `RandomSampler` class that will generate indices randomly with replacement
+
+        Example
+        -------
+        >>> sampler = Sampler.random_with_replacement(5)
+        >>> print(sampler.get_samples(10))
+        [3 4 0 0 2 3 3 2 2 1]
+        >>> print(next(sampler))
+        3
+        >>> print(sampler.next())
+        4
+
+        >>> sampler = Sampler.random_with_replacement(num_indices=4, prob=[0.7,0.1,0.1,0.1])
+        >>> print(sampler.get_samples(10))
+        [0 1 3 0 0 3 0 0 0 0]
+        """
+
+        sampler = SamplerRandom(
+            num_indices=num_indices,
+            sampling_type='random_with_replacement',
+            prob=prob,
+            replace=True,
+            seed=seed
+        )
+        return sampler
+
+    @staticmethod
+    def random_without_replacement(num_indices, seed=None):
+        """
+        Instantiates a sampler which outputs an index between 0 - num_indices. Once sampled the index will not be sampled again until all indices have been returned.
+
+        Parameters
+        ----------
+        num_indices: int
+            The sampler will select from a range of indices 0 to num_indices.
+
+        seed: int, optional
+            Used to initialise the random number generator where repeatability is required.  
+
+        Returns  
+        -------  
+        `RandomSampler`  
+            An instance of the `RandomSampler` class that will generate indices randomly without replacement 
+
+        Example
+        -------
+        >>> sampler=Sampler.randomWithoutReplacement(num_indices=7, seed=1)
+        >>> print(sampler.get_samples(16))
+        [6 2 1 0 4 3 5 1 0 4 2 5 6 3 3 2]
+
+        """
+
+        sampler = SamplerRandom(
+            num_indices=num_indices,
+            sampling_type='random_without_replacement',
+            replace=False,
+            seed=seed
+        )
+        return sampler
+
+    @staticmethod
+    def from_function(num_indices, function, prob_weights=None):
+        """
+        Instantiate a sampler that wraps a function for index selection.
+
+        Parameters
+        ----------
+        num_indices: int
+            The sampler will select from a range of indices 0 to num_indices.
+
+        function : Callable[[int], int]
+            A deterministic function that takes an integer iteration number and returns an integer between 0 and num_indices.
```

Comment:
```suggestion
    function : callable
        A deterministic function that takes an integer as an argument, representing the iteration number, and returns an integer between 0 and num_indices. The function signature should be function(iteration_number: int) -> int
```

Maybe this is a bit clearer?